barretenberg/cpp/src/barretenberg/world_state/world_state.hpp

```diff
@@ -0,0 +1,378 @@
+#pragma once
+
+#include "barretenberg/crypto/merkle_tree/append_only_tree/append_only_tree.hpp"
+#include "barretenberg/crypto/merkle_tree/hash.hpp"
+#include "barretenberg/crypto/merkle_tree/hash_path.hpp"
+#include "barretenberg/crypto/merkle_tree/indexed_tree/indexed_leaf.hpp"
+#include "barretenberg/crypto/merkle_tree/indexed_tree/indexed_tree.hpp"
+#include "barretenberg/crypto/merkle_tree/lmdb_store/functions.hpp"
+#include "barretenberg/crypto/merkle_tree/lmdb_store/lmdb_environment.hpp"
+#include "barretenberg/crypto/merkle_tree/lmdb_store/lmdb_store.hpp"
+#include "barretenberg/crypto/merkle_tree/node_store/cached_tree_store.hpp"
+#include "barretenberg/crypto/merkle_tree/node_store/tree_meta.hpp"
+#include "barretenberg/crypto/merkle_tree/response.hpp"
+#include "barretenberg/crypto/merkle_tree/signal.hpp"
+#include "barretenberg/crypto/merkle_tree/types.hpp"
+#include "barretenberg/ecc/curves/bn254/fr.hpp"
+#include "barretenberg/serialize/msgpack.hpp"
+#include "barretenberg/world_state/history.hpp"
+#include "barretenberg/world_state/struct.hpp"
+#include "barretenberg/world_state/tree_with_store.hpp"
+#include <algorithm>
+#include <exception>
+#include <iterator>
+#include <optional>
+#include <stdexcept>
+#include <type_traits>
+#include <unordered_map>
+#include <variant>
+
+namespace bb::world_state {
+
+using crypto::merkle_tree::index_t;
+
+using HashPolicy = crypto::merkle_tree::PedersenHashPolicy;
+
+using FrStore = crypto::merkle_tree::CachedTreeStore<crypto::merkle_tree::LMDBStore, fr>;
+using FrTree = crypto::merkle_tree::AppendOnlyTree<FrStore, HashPolicy>;
+
+using NullifierStore =
+    crypto::merkle_tree::CachedTreeStore<crypto::merkle_tree::LMDBStore, crypto::merkle_tree::NullifierLeafValue>;
+using NullifierTree = crypto::merkle_tree::IndexedTree<NullifierStore, HashPolicy>;
+
+using PublicDataStore =
+    crypto::merkle_tree::CachedTreeStore<crypto::merkle_tree::LMDBStore, crypto::merkle_tree::PublicDataLeafValue>;
+using PublicDataTree = crypto::merkle_tree::IndexedTree<PublicDataStore, HashPolicy>;
+
+using Tree = std::variant<TreeWithStore<FrTree>, TreeWithStore<NullifierTree>, TreeWithStore<PublicDataTree>>;
```

Comment:
Variants are not considered kosher/idiomatic in C++. This doesn't feel like a typescript `type Tree = A | B | C`. In C++ it feels more like a casting to any and then casting to some type and hoping.

What's usually done is to set up a common interface and use that.

I'm not saying you change it because it might be quite difficult. I'm only mentioning that coming from a C++ bg it doesn't feel right unless there's no other option.

Reply:
Indeed, it's difficult to have a common interface in this case (actually the `IndexedTree` does already extend `AppenyOnlyTree`). The other thing I had in mind was to just not use a map/array and instead have each tree be a member variable which mean having a switch or if/else chain in each function.

I haven't touched C++ in a long time (but modern C++ looks awesome!), my understanding of variant is it just takes away the switch statement and has the compiler "write" it?